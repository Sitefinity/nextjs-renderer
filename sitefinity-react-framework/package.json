{
  "name": "sitefinity-react-framework",
  "version": "0.1.0",
  "private": true,
  "publishConfig": {
    "access": "public"
  },
  "license": "See license in EULA.md",
  "dependencies": {
    "@types/react": "18.2.28",
    "react": "18.2.0",
    "tslib": "2.3.1",
    "typescript": "5.2.2"
  },
  "scripts": {
    "build": "tsc"
  }
}
