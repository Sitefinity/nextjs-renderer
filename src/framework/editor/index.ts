export * from './renderer-contract-interfaces';
export * from './services';
export * from './utils';
export * from './widgets';

