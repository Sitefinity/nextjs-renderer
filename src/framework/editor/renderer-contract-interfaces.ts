import { WidgetModel } from "../widgets/widget-model";

export interface RendererContract {
    getWidgetMetadata(args: GetWidgetMetadataArgs): Promise<ComponentMetadata>;
    renderWidget(args: RenderWidgetArgs): Promise<RenderResult>;
    getCategories(args: GetCategoriesArgs): Promise<Array<string>>;
    getWidgets(args: GetWidgetsArgs): Promise<TotalCountResult<Array<WidgetSection>>>;
}

export interface TotalCountResult<T> {
    dataItems: T;
    totalCount: number;
}

export interface RenderResult {
    element: HTMLElement,
    content: string,
    scripts: Array<Script>
}

export interface Script {
    src: string;
    id: string;
}

export interface WidgetItem {
    name: string;
    title?: string;
    addWidgetTitle?: string;
    addWidgetName?: string;
    thumbnailUrl?: string;
    initialProperties?: Array<{ name: string, value: string}>
}

export interface WidgetSection {
    title: string;
    widgets: WidgetItem[];
}

export interface RenderWidgetArgs {
    dataItem: DataItem;
    siteId: string;
    model: WidgetModel<any>;
    token?: Token;
}
export interface GetCategoriesArgs {
    token?: Token;
    toolbox?: string;
}

export interface GetWidgetsArgs {
    dataItem: DataItem;
    category: string;
    search: string;
    skip: number;
    take: number;
    token?: Token;
    toolbox?: string;
}

export interface GetWidgetMetadataArgs {
    dataItem: DataItem;
    widgetKey: string;
    widgetName: string;
    siteId: string;
    token?: Token;
}

export interface ComponentMetadata {
    Caption: string,
    Name: string,
    PropertyMetadata: SectionGroup[],
    PropertyMetadataFlat: PropertyMetadata[],
}

export interface PropertyMetadata {
    Name: string;
    Title: string;
    Type: string;
    DefaultValue: string;
    Properties: { [key: string]: any }
}

export interface SectionGroup {
    Name: string;
    Sections: SectionData[];
}

export interface SectionData {
    Name: string,
    Title: string
    Properties: Array<PropertyMetadata>
}

export interface Token {
    type: string,
    value: string
}

export interface DataItem {
    readonly provider: string;
    readonly culture: string;
    readonly key: string;
}
