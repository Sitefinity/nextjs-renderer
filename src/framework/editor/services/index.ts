export * from './detail-item';
export * from './render-widget-service';
export * from './request-context';
