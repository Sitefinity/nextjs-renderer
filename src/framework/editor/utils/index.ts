export * from './classNames';
export * from './getUniqueId';
export * from './guid';
export * from './rest-extensions';
