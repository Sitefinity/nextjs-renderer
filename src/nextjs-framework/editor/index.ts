export * from './renderer-contract-interfaces';
export * from './utils/classNames';
export * from './utils/getUniqueId';
export * from './utils/guid';

export * from './detail-item';
export * from './request-context';

export * from './widget-framework/attributes';
export * from './widget-framework/content-list-settings';
export * from './widget-framework/link-model';
export * from './widget-framework/mixed-content-context';
export * from './widget-framework/widget-context';
export * from './widget-framework/widget-editor-metadata';
export * from './widget-framework/widget-metadata';
export * from './widget-framework/widget-model';
export * from './widget-framework/widget-registry';
