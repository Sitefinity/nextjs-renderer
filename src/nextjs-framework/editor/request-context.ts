import { PageLayoutServiceResponse, SdkItem } from '../rest-sdk';
import { DetailItem } from './detail-item';

export interface RequestContext {
    layout: PageLayoutServiceResponse;
    searchParams: { [key: string]: string; };
    detailItem?: DetailItem;
    culture: string;
    isEdit: boolean;
    isPreview: boolean;
    isLive: boolean;
}
