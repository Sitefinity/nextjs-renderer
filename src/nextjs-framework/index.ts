export * from './pages/render-page';
export * from './pages/render-widget';
export * from './pages/utils';
