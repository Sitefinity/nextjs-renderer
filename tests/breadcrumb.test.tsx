
import { render, waitFor } from '@testing-library/react';
import { MockedBreadcrumbRestService } from './mocks/mock-breadcrumb.service';
import { Breadcrumb, BreadcrumbEntity } from '@progress/sitefinity-react-framework';
import { RequestContext, WidgetModel } from '../src/nextjs-framework/editor';

it('should render breadcrumb', async () => {
    const callToAction = await Breadcrumb({
        model,
        requestContext,
        metadata,
        restService: MockedBreadcrumbRestService });
    const { container } = render(callToAction);

    await waitFor(() => {
        expect(container).toMatchSnapshot();
    });
});

const metadata = {
    designerMetadata: {
        Name: 'SitefinityBreadcrumb',
        Caption: 'Breadcrumb',
        PropertyMetadata: [Array]
    },
    componentType: [Breadcrumb],
    editorMetadata: { Title: 'Breadcrumb' },
    ssr: true
};

const model: WidgetModel<BreadcrumbEntity> = {
    Id: 'c53ba795-8c6c-4b4e-a5f3-531e8f8e8122',
    Name: 'SitefinityBreadcrumb',
    PlaceHolder: 'Body',
    Caption: 'Breadcrumb',
    ViewName: '',
    Lazy: false,
    Properties: {},
    Children: []
};

const requestContext: RequestContext = {
    pageNode: {
        Culture: 'en',
        SiteId: '4c922118-f076-4e24-9193-93e004f50107',
        MetadataHash: '-1129976431',
        Id: '831706cd-10e1-4184-b2e5-ef9f9ad0fdb5',
        TemplateName: 'Default',
        Renderer: 'React',
        UrlParameters: [],
        HasVariations: false,
        Site: {
            Id: '4c922118-f076-4e24-9193-93e004f50107',
            Name: 'Quantum International',
            DefaultCulture: 'en',
            Cultures: [Array],
            TimeZone: 'UTC;0;UTC;UTC;UTC;;',
            IsSubFolder: false
        },
        User: null,
        ComponentContext: {
            HasLazyComponents: false,
            OrphanedControls: [],
            Components: [Array],
            Fields: null,
            Site: null,
            User: null
        },
        Scripts: [],
        DetailItem: null,
        MetaInfo: {
            Title: 'integration-test',
            Description: '',
            HtmlInHeadTag: null,
            OpenGraphTitle: 'integration-test',
            OpenGraphDescription: '',
            OpenGraphImage: null,
            OpenGraphVideo: null,
            OpenGraphType: null,
            OpenGraphSite: null,
            CanonicalUrl: 'https://localhost:5001/integration-test'
        },
        Fields: {
            Title: 'integration-test',
            UrlName: 'integration-test',
            LastModified: '2023-11-13T13:13:09.183Z',
            DateCreated: '2023-11-13T13:13:09.183Z',
            Crawlable: true,
            ParentId: 'f669d9a7-009d-4d83-ddaa-000000000002',
            RootId: 'f669d9a7-009d-4d83-ddaa-000000000002',
            IsHomePage: false,
            'AvailableLanguages@odata.type': '#Collection(String)',
            AvailableLanguages: [Array],
            ViewUrl: '/integration-test'
        }
    },
    searchParams: {
        sf_site: '4c922118-f076-4e24-9193-93e004f50107',
        sfaction: 'edit',
        sf_version: '11'
    },
    detailItem: null,
    culture: 'en',
    isEdit: true,
    isPreview: false,
    isLive: false
};
