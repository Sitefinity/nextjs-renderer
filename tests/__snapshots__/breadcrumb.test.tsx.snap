// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render breadcrumb with empty model 1`] = `
<div>
  <div
    class=""
    data-sfhasquickeditoperation="true"
    data-sfid="dbd6fe0e-eaea-426a-bab6-8689ea755a39"
    data-sfiscontentwidget="true"
    data-sfisempty="false"
    data-sfisemptyvisualhidden="false"
    data-sfisorphaned="false"
    data-sfname="SitefinityBreadcrumb"
    data-sftitle="Breadcrumb"
    draggable="true"
  >
    Breadcrumb is visible when you are on a particular page.
  </div>
</div>
`;

exports[`should render breadcrumb with full path to current page 1`] = `
<div>
  <div
    class=""
    data-sfhasquickeditoperation="true"
    data-sfid="c53ba795-8c6c-4b4e-a5f3-531e8f8e8122"
    data-sfiscontentwidget="true"
    data-sfisempty="false"
    data-sfisemptyvisualhidden="false"
    data-sfisorphaned="false"
    data-sfname="SitefinityBreadcrumb"
    data-sftitle="Breadcrumb"
    draggable="true"
  >
    Breadcrumb is visible when you are on a particular page.
  </div>
</div>
`;

exports[`should render breadcrumb with margins 1`] = `
<div>
  <div
    class="mb-4 ms-5 me-3"
    data-sfhasquickeditoperation="true"
    data-sfid="52a11ccb-95b7-4b99-b09d-7e4ccf24668d"
    data-sfiscontentwidget="true"
    data-sfisempty="false"
    data-sfisemptyvisualhidden="false"
    data-sfisorphaned="false"
    data-sfname="SitefinityBreadcrumb"
    data-sftitle="Breadcrumb"
    draggable="true"
  >
    <nav
      aria-label="Full path to the current page"
    >
      <ol
        class="breadcrumb"
      >
        <li
          class="breadcrumb-item"
        >
          <a
            href="/home"
          >
            Home
          </a>
        </li>
        <li
          aria-current="page"
          class="breadcrumb-item active"
        >
          react new
        </li>
      </ol>
    </nav>
  </div>
</div>
`;
